cvc5/Solver.lean

```diff
@@ -314,8 +385,77 @@ private opaque mkIntegerFromString : TermManager → String → Term
 def mkInteger (tm : TermManager) : Int → Term :=
   (mkIntegerFromString tm) ∘ toString
 
+/-- Create operator of Kind:
+
+- `Kind.BITVECTOR_EXTRACT`
+- `Kind.BITVECTOR_REPEAT`
+- `Kind.BITVECTOR_ROTATE_LEFT`
+- `Kind.BITVECTOR_ROTATE_RIGHT`
+- `Kind.BITVECTOR_SIGN_EXTEND`
+- `Kind.BITVECTOR_ZERO_EXTEND`
+- `Kind.DIVISIBLE`
+- `Kind.FLOATINGPOINT_TO_FP_FROM_FP`
+- `Kind.FLOATINGPOINT_TO_FP_FROM_IEEE_BV`
+- `Kind.FLOATINGPOINT_TO_FP_FROM_REAL`
+- `Kind.FLOATINGPOINT_TO_FP_FROM_SBV`
+- `Kind.FLOATINGPOINT_TO_FP_FROM_UBV`
+- `Kind.FLOATINGPOINT_TO_SBV`
+- `Kind.FLOATINGPOINT_TO_UBV`
+- `Kind.INT_TO_BITVECTOR`
+- `Kind.TUPLE_PROJECT`
+
+See `cvc5.Kind` for a description of the parameters.
+
+- `kind` The kind of the operator.
+- `args` The arguments (indices) of the operator.
+
+If `args` is empty, the `Op` simply wraps the `cvc5.Kind`. The `Kind` can be used in `Solver.mkTerm`
+directly without creating an `Op` first.
+-/
+@[extern "termManager_mkOpOfIndices"]
+opaque mkOpOfIndices : TermManager → (kind : Kind) → (args : Array Nat) → Op
+
+/-- Create operator of kind:
+
+- `Kind.DIVISIBLE` (to support arbitrary precision integers)
+
+See `cvc5.Kind` for a description of the parameters.
+
+- `kind` The kind of the operator.
+- `arg` The string argument to this operator.
+
+-/
+@[extern "termManager_mkOpOfString"]
+opaque mkOpOfString : TermManager → (kind : Kind) → (arg : String) → Op

```

Comment:
You can replace `(arg : String)` with `(arg : Int)` since Lean has arbitrary precision integers. Look at the implementation of `mkInteger` for inspiration!

Reply:
Turns out this function (currently) accepts `Kind.DIVISIBLE` with a strictly positive argument. I modified it in the way you asked, but are we sure we want to have these `mkOp`-like functions?

Reply:
Unfortunately, we do...

Reply:
Okay, I ended up exposing the (currently) only safe way to use it (according to the doc):

```lean
@[extern "termManager_mkOpOfString"]
private opaque mkOpOfString : TermManager → (kind : Kind) → (arg : String) → Op

/-- Create divisibility-by operator, supports arbitrary precision. -/
def mkOpDivisible (tm : TermManager) (n : Nat) (_valid : 0 < n := by simp) : Op :=
  tm.mkOpOfString Kind.DIVISIBLE (toString n)
```

I'll let you resolve this conversation if you're fine with this.